src/format.rs

```diff
@@ -0,0 +1,888 @@
+                        (b'U', Spec::WeekNumberFromSunday),
+                        (b'V', Spec::WeekNumberIso8601),
+                        (b'W', Spec::WeekNumberFromMonday),
+                        (b'X', Spec::CombinationTime24h),
+                        (b'Y', Spec::Year4Digits),
+                        (b'Z', Spec::TimeZoneName),
+                        (b'a', Spec::WeekDayNameAbbr),
+                        (b'b', Spec::MonthNameAbbr),
+                        (b'c', Spec::CombinationDateTime),
+                        (b'd', Spec::MonthDayZero),
+                        (b'e', Spec::MonthDaySpace),
+                        (b'g', Spec::YearIso8601Rem100),
+                        (b'h', Spec::MonthNameAbbr),
+                        (b'j', Spec::YearDay),
+                        (b'k', Spec::Hour24hSpace),
+                        (b'l', Spec::Hour12hSpace),
+                        (b'm', Spec::Month),
+                        (b'n', Spec::Newline),
+                        (b'p', Spec::MeridianUpper),
+                        (b'r', Spec::CombinationTime12h),
+                        (b's', Spec::SecondsSinceEpoch),
+                        (b't', Spec::Tabulation),
+                        (b'u', Spec::WeekDayFrom1),
+                        (b'v', Spec::CombinationVmsDate),
+                        (b'w', Spec::WeekDayFrom0),
+                        (b'x', Spec::CombinationDate),
+                        (b'y', Spec::YearRem100),
+                        (b'z', Spec::TimeZoneOffsetHourMinute),
+                    ];
+                    assert_sorted_spec(possible_specs);
+                    possible_specs
+                };
+
+                match cursor.next() {
+                    Some(x) => match POSSIBLE_SPECS.binary_search_by_key(&x, |&(c, _)| c) {
+                        Ok(index) => Some(POSSIBLE_SPECS[index].1),
+                        Err(_) => None,
+                    },
+                    None => return Err(FormatError::InvalidFormat),
+                }
+            } else if cursor.read_optional_tag(b"z") {
+                match colons.len() {
+                    1 => Some(Spec::TimeZoneOffsetHourMinuteColon),
+                    2 => Some(Spec::TimeZoneOffsetHourMinuteSecondColon),
+                    3 => Some(Spec::TimeZoneOffsetColonMinimal),
+                    _ => None,
+                }
+            } else {
+                None
+            };
+
+            if let Some(spec) = spec {
+                Piece::new(width, padding, has_left_padding, case, spec).fmt(&mut f, self.time)?;
+            } else {
+                // No valid format spec found
+                let remaining_after = cursor.remaining();
+                let text = &remaining_before[..remaining_before.len() - remaining_after.len()];
+                f.write_all(text)?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+fn year_width(year: i32) -> usize {
```

Comment:
my naive reaction when reviewing this function is that it could be replaced with `year.log10() + 1`, but that won't handle the cases where year is zero or negative.

I'm not sure if the impl with `log10` would end up being more compact so it might not be worth it.

maybe?

```rust
fn year_width(year: i32) -> u32 {
    match year {
        0 => 1,
        y if y < 0 => y.checked_neg().unwrap_or(i32::MAX).log10() + 2, // add an extra width for the `-` sign
        y => y.log10() + 1,
    }
}
```

Reply:
`log10` for integers are unstable as of Rust 1.62: https://doc.rust-lang.org/std/primitive.i32.html#method.log10.

Reply:
ahh gotcha, I missed that in the docs.